File: transcode-go/middleware/builtin/recovery.go
package builtin

// recovery.go records a panic that the transcoder caught.

import (
	"log/slog"

	"github.com/the-protobuf-project/http/transcode-go/middleware"
)

// Recovery records a panic as a completed call.
//
// A panic in one handler must not take down the connection, because on HTTP/2
// that connection is carrying other people's requests. It also must not reach
// the client: a panic value frequently contains a file path, a slice index, or
// a fragment of the data that caused it.
//
// The catching itself happens in the transcoder, which is the only thing that owns
// the call — an interceptor cannot recover from a panic in a function it never
// called. This interceptor is what makes the panic visible to the same
// observability stack as everything else, rather than only in the transcoder's own
// logger.
type Recovery struct {
	// logger receives the record.
	logger *slog.Logger
}

// NewRecovery returns the interceptor.
func NewRecovery(logger *slog.Logger) *Recovery {
	if logger == nil {
		logger = slog.Default()
	}
	return &Recovery{logger: logger}
}

// Name implements [middleware.Interceptor].
func (*Recovery) Name() string { return "recovery" }

// OnComplete records a call that ended in GATEWAY_PANIC.
func (r *Recovery) OnComplete(cx *middleware.CallCx, outcome middleware.Outcome) {
	if !outcome.Failed() {
		return
	}
	info := outcome.Err.ErrorInfo()
	if info == nil || info.Reason != "GATEWAY_PANIC" {
		return
	}

	r.logger.Error("handler panicked",
		"method", cx.Method.FullName,
		"template", cx.Template,
		"elapsedMs", cx.Elapsed().Milliseconds(),
	)
}

File: http-rs/grpc-http/src/middleware/builtin/recovery.rs
//! Turning a panic into a response.

use crate::error::GatewayError;
use crate::middleware::{CallCx, Interceptor, Outcome};

/// Catches an unwind and renders it as `500`.
///
/// A panic in one handler must not take down the connection, because on HTTP/2
/// and HTTP/3 that connection is carrying other people's requests. It also must
/// not reach the client: a panic payload frequently contains a file path, a
/// slice index, or a fragment of the data that caused it.
///
/// This interceptor records the panic. The catching itself happens in the
/// generated handler, which is the only place that owns the call — an
/// interceptor cannot wrap a future it never sees.
#[derive(Debug, Clone, Copy, Default)]
pub struct Recovery;

impl Recovery {
    /// Builds the interceptor.
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Interceptor for Recovery {
    fn name(&self) -> &'static str {
        "recovery"
    }

    /// Logs a caught panic at error level, with the method that caused it.
    fn on_complete(&self, cx: &CallCx<'_>, outcome: &Outcome<'_>) {
        if let Outcome::Failure(err) = outcome
            && is_panic(err)
        {
            tracing::error!(
                method = cx.route.method,
                template = cx.route.template,
                elapsed_ms = cx.elapsed().as_millis(),
                "handler panicked; returned 500"
            );
        }
    }
}

/// Whether an error came from a caught panic.
fn is_panic(err: &GatewayError) -> bool {
    err.details.iter().any(|detail| {
        matches!(detail, crate::error::Detail::ErrorInfo(info) if info.reason == "GATEWAY_PANIC")
    })
}
